iocore/net/quic/QUICConnectionTable.cc

```diff
@@ -26,21 +26,25 @@
 int
 QUICConnectionTable::insert(QUICConnectionId cid, QUICConnection *connection)
 {
-  this->_connections.put(cid, connection);
-  // if (this->_cids.get(connection->endpoint()) == nullptr) {
-  //     this->_cids.put(connection->endpoint(), cid);
-  // }
+  SCOPED_MUTEX_LOCK(lock, _connections.lock_for_key(cid), this_ethread());
+  QUICConnection *ret_connection = _connections.insert_entry(cid, connection);
+  // To check whether the return value is nullptr in case memory leak.
+  // The return value isn't nullptr, the new value will take up the slot and return old value.
+  if (ret_connection) {
+    // TODO: free the old connection.
```

Comment:
This shouldn't happen, right? Maybe we should return an error code instead of silently freeing it.

Reply:
Yes, It shouldn't happen(CID is unique value).
If the CID has duplicate value, the old value was be returned, it should be freed.

Reply:
The connection have to be freed somewhere but I think here is not the appropriate place because this table is just a container. Some another class should be responsible for freeing connections. Probably, the class that allocate/instantiate the connection.

How about escalating the odd situation to the caller side with an error code and freeing the connection on the caller side? Then we can do something more (Decrement/Increment some stats, Output log with other information).

Reply:
Fine, I can provide the pointer of QUICConnection as returning value in insert functions, like remove functions. To determine if the returning value is nullptr by caller side. If the returning value isn't nullptr, it should be freed by caller.

Reply:
@JordyLiu26 maybe return a smart pointer since we are on modern c++?

Reply:
Do you have any ideas or plans for looking up a connection with 5-tuple?
I was going to return a integer to indicate what key we can use (and errors).

For example:
Negative numbers: errors
0: CID (a peer need to send CID)
1: 5-tuple (CID is available too, but a peer can omit it)

Reply:
Since the caller has a CID, the caller can `lookup()` the connection, `erase()` it from a table and free it, after a failure of `insert()`.

Reply:
I think the connection table is a container, it use an integer as a key in HashTable. The ConnectionTable is only concerned about key (integer). The ConnectionTable can provide make_key functions to convert  5-tuple or CID to key.

Reply:
> The ConnectionTable can provide make_key functions to convert 5-tuple or CID to key.

Why don't you automate the conversion? Keys for internal hashtable (MT_hashtable) have to be integer but ConnectionTable doesn't have to be the same. As you can see, currently `lookup()` receives a QUICPacket and a 5-tuple. So you can internally extract a integer key (CID) from them if necessary.

A return value from `insert()` would be used only once to tell a client whether CID is omittable when we send TransportParameters. After that, the caller of `lookup()` don't need to care about how keys are derived, and it can simply do the same thing on every packets. , The commented out (incomplete) code does this trick.

Reply:
Sorry, I did not notice the comments. 
// Map<QUICFiveTuple, QUICConnectionId> _cids;
There are two tables in ConnectionTable(CID and 5-tuples ). As you mentioned, to return a integer to indicate what key we can use.

If the container have tow hash_table, it becomes more complex and inefficiently. To get two locks when insert/erase/lookup a value to/from two tables. I think one table is better. The CID and 5-tuples store in the same table. 
The CID is long integer. for example: 15532264403001802089.
5-tuples is integer. src.port() + dst.port() + protocol;  for example: 32587. (Maybe generate the hash code by ats_ip_port_hash)
So, the key is different, there is no hash collision in hash table. To find the VC by 5-tuples if there is no CID in the packet. The key point is two keys can not be the same in the same table.

Reply:
I agree. It would be great if we could use one hashtable for the both.

However, CID can be 32587 too. So we need something to avoid the collision.

Reply:
void
  randomize()
  {
    std::random_device rnd;
    this->_id = (static_cast<uint64_t>(rnd()) << 32) + rnd();
  };
The rnd.min() is 0, rnd.max() is 4294967295 (unsigned int), but the rnd was shifted left 32 bits. So the CID should greater than 4294967296.(except 0)

Reply:
So if `(static_cast<uint64_t>(rnd()) << 32)` was 0, CID will be less than 4294967295. And more importantly, every 64 bits numbers should be valid CIDs. Using only 4294967296 to UINT64_MAX would be a kind of protocol violation.

Reply:
Maybe, modify the following,

void
randomize()
{
  std::random_device rnd;
  std::mt19937 gen(rnd); // to generate a single random number that is used as a seed for a pseudorandom number generator (Mersenne Twister)
  std::uniform_int_distribution<> dis(min, max);  // set min and max number. [min, max]
  this->_id = (static_cast<uint64_t>(dis(gen)) << 32) + rnd();
};

Reply:
If I understand correctly, it would ensure that upper 32 bits are not 0. However that's not what I really wanted to say. Probably, we need two hashtables unless a hashtable supports longer keys (maybe 224 bits at least).

There are tow reasons:
a) CID is unsigned 64 bits integer. Reserving some bits to not conflict with 5-tuple reduces actual ID range available.
b) Imagine IPv6 support. 5-tuple is obviously longer than CID. Even if we use only source address and source port, technically, we need 224 (64 + 128 + 32) bits space.

Anyway, the specification is not fixed, so the table doesn't need to be perfect. Let's revisit when we support omitting CID.

Reply:
Yes. If one table contains CID and 5-tuples, the key was divided two parts to ensure the table has unique key.